### PR TITLE
Fix failing tuist install when a package is removed

### PR DESCRIPTION
**Pull request by fortmarek**
Resolves https://github.com/tuist/tuist/issues/5489

### Short description 📝

If you remove a dependency from your `Package.swift` and run `tuist install`, you will likely get an error. This is because `swift package resolve` does a full clone of a repo and `tuist generate` generates derived files into the full checkout of a dependency. This means we're adding new, unstaged files into the git repo of a dependency and `swift package resolve` fails because it refuses to delete a locally checked out repo with changes.

The solution to this is to generate all derived files and the actual `.xcodeproj` files into a new directory (I've chosen `.build/tuist-derived`).

### How to test the changes locally 🧐

Generate `app_with_spm_dependencies`, remove a dependency, and run `tuist install`.

After generating `app_with_spm_dependencies`, no checked out SPM repository should have unstaged changes. 

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint:fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [x] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/TuistLoader/SwiftPackageManager/PackageInfoMapper.swift [hunk 1]
```diff
@@ -906,16 +907,11 @@ extension ProjectDescription.TargetDependency {
 }
 
 extension ProjectDescription.Headers {
-    fileprivate static func from(moduleMap: ModuleMap?, publicHeadersPath: AbsolutePath) throws -> Self? {
+    fileprivate static func from(moduleMap: ModuleMap?, publicHeadersPath _: AbsolutePath) throws -> Self? {
```
Comment (danyf90) [inline Sources/TuistLoader/SwiftPackageManager/PackageInfoMapper.swift:910]: If it's not used we can remove the `publicHeadersPath` parameter, right?

File: Sources/TuistDependencies/Mappers/ExternalDependencyPathWorkspaceMapper.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            mappedProjects.flatMap(\.1)
+        )
+    }
+
+    // MARK: - Helpers
+
+    private func map(project: Project) throws -> (Project, [SideEffectDescriptor]) {
+        guard project.isExternal,
+              // We don't want to update local packages (which are defined outside the `checkouts` directory in `.build`
+              project.path.parentDirectory.parentDirectory.basename == Constants.SwiftPackageManager.packageBuildDirectoryName
```
Comment (danyf90) [inline Sources/TuistDependencies/Mappers/ExternalDependencyPathWorkspaceMapper.swift:25]: We are computing the path from the folder where the directory is located, right?

Wouldn't be simpler (and more future proof) to calculate the path from the Tuist root (i.e. where the `Tuist` folder is located)?
I would put things in `Tuist/Dependencies/Derived/<Project>/<Target>` or something similar

**Review by danyf90 — APPROVED** [review]

File: Sources/TuistDependencies/Mappers/ExternalDependencyPathWorkspaceMapper.swift (hunk 2 above)
Comment (fortmarek) [reply Sources/TuistDependencies/Mappers/ExternalDependencyPathWorkspaceMapper.swift:25]: that would be another dir that users need to gitignore, unnecessarily so. I don't think the inner structure of .build will change and I think it's good to have the derived directory close to the package sources.

Comment (danyf90) [reply Sources/TuistDependencies/Mappers/ExternalDependencyPathWorkspaceMapper.swift:25]: They should git ignore the whole Tuist/Dependencies already, isn't it?

Comment (fortmarek) [reply Sources/TuistDependencies/Mappers/ExternalDependencyPathWorkspaceMapper.swift:25]: But that directory doesn't exist anymore from tuist 4

Comment (fortmarek) [reply Sources/TuistDependencies/Mappers/ExternalDependencyPathWorkspaceMapper.swift:25]: And I don't get what's the benefit of creating another directory when `.build` should already be gitignored and contains everything related to the SPM dependencies 🤷‍♂️

**Pull request merged**